---
title: "Integration analysis of scRNA-seq"
author: "Pang-Kuo Lo"
date: "11/5/2019"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Integration Analysis of Single-Cell RNA-seq Datasets

The "Seurat" library package developed by Satija Lab is a useful R-based tool for single-cell genomics analysis. Seurat is used here for integration analysis of single-cell RNA-seq datasets from two different BMMC samples (aml027 and aml035) with two different conditions (pre-transplantation and post-transplantation). These four datasets can be downloaded from the resources of the 10X genomics website.  


```{r C1, message=FALSE, warning=FALSE}
## Load required library packages
library(Seurat)
library(cowplot)
library(ggplot2)
library(rafalib)
```


The first step is to load 10X datasets to create their respective Seurat objects using the Read10X() function syntax. Each single-cell RNA-seq (scRNA-seq) dataset contain three files, including barcodes.tsv, genes.tsv and matrix.mtx. These three files need to be placed in the same folder for being read by Read10X(). These files need to be decompressed before they can be read if they are compressed as tar or gz.

```{r C2, message=FALSE, warning=FALSE}
AML027_pre.data <- Read10X(data.dir="/Users/Pang-Kuo/Desktop/NGS_ML_Analysis/single_cell_RNA-seq_analysis/AML027_Pre-transplant_BMMCs/hg19", gene.column = 2)
AML027_post.data <- Read10X(data.dir="/Users/Pang-Kuo/Desktop/NGS_ML_Analysis/single_cell_RNA-seq_analysis/AML027_Post-transplant_BMMCs/hg19", gene.column = 2)
AML035_pre.data <- Read10X(data.dir="/Users/Pang-Kuo/Desktop/NGS_ML_Analysis/single_cell_RNA-seq_analysis/AML035_Pre-transplant_BMMCs/hg19", gene.column = 2)
AML035_post.data <- Read10X(data.dir="/Users/Pang-Kuo/Desktop/NGS_ML_Analysis/single_cell_RNA-seq_analysis/AML035_Post-transplant_BMMCs/hg19", gene.column = 2)
```


The second step is to create their respective Seurat objects from their matrix data.

```{r C3, message=FALSE, warning=FALSE}
# Create a function for converting matrix data into a Seurat object
Seurat_object <- function(matrix, object_name) {
  object <- CreateSeuratObject(counts = matrix, project = object_name, min.cells = 5)
  object$INT <- object_name
  object <- subset(object, subset = nFeature_RNA > 500)
  object <- NormalizeData(object, verbose = FALSE)
  object <- FindVariableFeatures(object, selection.method = "vst", nfeatures = 2000)
  return(object)
}

# Use the Seurat_object function to create Seurat objects
AML027_pre <- Seurat_object(AML027_pre.data, "AML027_pre")
AML027_post <- Seurat_object(AML027_post.data, "AML027_post")
AML035_pre <- Seurat_object(AML035_pre.data, "AML035_pre")
AML035_post <- Seurat_object(AML035_post.data, "AML035_post")
```


The third step is to perform integration for creating an integrated Seurat object.

```{r C4, message=FALSE, warning=FALSE}
#Perform integration
AML.anchors <- FindIntegrationAnchors(object.list = list(AML027_pre, AML027_post, AML035_pre, AML035_post), dims = 1:15)
AML.combined <- IntegrateData(anchorset = AML.anchors, dims = 1:15)
```


The fourth step is to scale and center the data and perform a PCA analysis.

```{r C5, message=FALSE, warning=FALSE}
DefaultAssay(AML.combined) <- "integrated"
AML.combined <- ScaleData(AML.combined, verbose = FALSE)
AML.combined <- RunPCA(AML.combined, npcs = 30, verbose = FALSE)
```


PCA plots can be created using the DimPlot function syntax.

```{r C6, echo=TRUE, fig.height=4.5, fig.width=9, message=FALSE, warning=FALSE}
DimPlot(AML.combined, reduction = "pca", split.by = "INT", pt.size = 0.8)
# Or
# PCAPlot(AML.combined, split.by = "INT", pt.size = 0.8)
```



The fifth step is to perform clustering.

```{r C7, message=FALSE, warning=FALSE}
AML.combined <- FindNeighbors(AML.combined, reduction = "pca", dims = 1:15)
AML.combined <- FindClusters(AML.combined, resolution = 0.5)
```


The sixth step is to perform t-SNE analysis.

```{r C8, message=FALSE, warning=FALSE}
AML.combined <- RunTSNE(AML.combined, reduction = "pca", dims = 1:15)
```


The t-SNE plots can be created using DimPlot or TSNEPlot function syntaxes.

```{r C9, echo=TRUE, fig.height=4.5, fig.width=9, message=FALSE, warning=FALSE}
DimPlot(AML.combined, reduction = "tsne", split.by = "INT", pt.size = 0.8)
# Or
# TSNEPlot(AML.combined, split.by = "INT", pt.size = 0.8)
```



The seventh step is to perform UMAP analysis.

```{r C10, message=FALSE, warning=FALSE}
AML.combined <- RunUMAP(AML.combined, reduction = "pca", dims = 1:15)
```


The UMAP plots can be created using DimPlot or UMAPPlot function syntaxes.

```{r C11, echo=TRUE, fig.height=4.5, fig.width=9, message=FALSE, warning=FALSE}
DimPlot(AML.combined, reduction = "umap", split.by = "INT", pt.size = 0.8)
# Or
# UMAPPlot(AML.combined, split.by = "INT", pt.size = 0.8)
```

The eighth step is to identify conserved cell-type-specific gene markers.

```{r C12, message=FALSE, warning=FALSE}
DefaultAssay(AML.combined) <- "RNA"
## Create an empty list for storing the output data of differentially expressed gene markers
gene_markers <- vector("list", 15)      
clust_num <- c(seq(1:15)-1)
names(gene_markers) <- paste0("c", c(seq(1:15)-1), "_markers")     

# Store differentially expressed genes in each cell cluster in the gene_markers list (Only cell clusters 0 - 4 are analyzed here as examples)
for (i in 1:5) {
  gene_markers[[i]] <- FindConservedMarkers(AML.combined, ident.1 = clust_num[i], grouping.var = "INT", verbose = FALSE)
}
```


Print the top 5 representative marker genes for cell clusters 0 - 4

```{r C13, echo=TRUE, message=FALSE, warning=FALSE}
for (i in 1:5) {
  print(names(gene_markers[i]));print(head(gene_markers[[i]],5)[,c(2,5,7,10,12,15,17,20)])
}
```


The FeaturePlot function syntax can be used to visualize expression patterns of cell-type-specific gene markers in UMAP plots (c0 cluster:"TYROBP", c1 cluster:"FAM178B", c2 cluster:"HBM", c3 cluster:"CD3D", c4 cluster:"IGLL1", c5 cluster:"RRM2")

```{r C14, echo=TRUE, fig.height=8, fig.width=7, message=FALSE, warning=FALSE}
FeaturePlot(AML.combined, features = c("TYROBP", "FAM178B","HBM", "CD3D", "IGLL1", "RRM2"), min.cutoff = "q9")
```

Check some well-known cell-type-specific marker (T cell marker: CD3D, B cell marker: CD79A)

```{r C15, echo=TRUE, fig.height=4, fig.width=7, message=FALSE, warning=FALSE}
FeaturePlot(AML.combined, features = c("CD3D", "CD79A"), min.cutoff = "q9")
```

To be easy visualization and comparison, identified cell clusters can be labeled in UMAP plots.

```{r C16, echo=TRUE, fig.height=4.5, fig.width=9, message=FALSE, warning=FALSE}
AML.combined <- RenameIdents(AML.combined, `0` = "C0", `1` = "C1", `2` = "C2", `3` ="C3-T cells", `4` = "C4-B cells", `5` = "C5", `6` = "C6", `7` = "C7", `8` = "C8", `9` = "C9", `10` = "C10", `11` = "C11", `12` = "C12", `13` = "C13", `14` = "C14")
                              
DimPlot(AML.combined, split.by = "INT", label = TRUE, pt.size = 0.8)
```


The AverageExpression syntax can be used to compute the average expression of genes in a specific cell cluster accross different conditions. Clusters 0 and 1 are analyzed here as examples.

```{r C17, message=FALSE, warning=FALSE}
C0_cluster <- subset(AML.combined, idents = "C0")
Idents(C0_cluster) <- "INT"
avg.C0_cluster <- log1p(AverageExpression(C0_cluster, verbose = FALSE)$RNA)
avg.C0_cluster$gene <- rownames(avg.C0_cluster)

C1_cluster <- subset(AML.combined, idents = "C1")
Idents(C1_cluster) <- "INT"
avg.C1_cluster <- log1p(AverageExpression(C1_cluster, verbose = FALSE)$RNA)
avg.C1_cluster$gene <- rownames(avg.C1_cluster)
```

Differentially expressed genes can be displayed in scatter plots using output data from above.

```{r C18, echo=TRUE, message=FALSE, warning=FALSE}
mypar(2,2)
with(avg.C0_cluster, plot(AML027_pre, AML027_post, pch=19, cex=0.7, main="Cell Cluster 0")) ; abline(0,1, lwd=2, col="red") 
with(avg.C0_cluster, plot(AML035_pre, AML035_post, pch=19, cex=0.7, main="Cell Cluster 0")) ; abline(0,1, lwd=2, col="red")              
with(avg.C1_cluster, plot(AML027_pre, AML027_post, pch=19, cex=0.7, main="Cell Cluster 1")) ; abline(0,1, lwd=2, col="red") 
with(avg.C1_cluster, plot(AML035_pre, AML035_post, pch=19, cex=0.7, main="Cell Cluster 1")) ; abline(0,1, lwd=2, col="red")              
```

The final step is to analyze differential expressed genes in a specific cell cluster/type across conditions using the FindMarkers syntax. Cell cluster C1 is analyzed here as an example. HSPA5 is identified as an upregulated gene in both AML027 and AML035 cell samples under the post-transplantation condition when compared to the pre-transplantation condition.

```{r C19, echo=TRUE, message=FALSE, warning=FALSE}
AML.combined$celltype.INT <- paste(Idents(AML.combined), AML.combined$INT, sep = "_")
AML.combined$celltype <- Idents(AML.combined)
AML.combined_2 <- AML.combined
Idents(AML.combined_2) <- "celltype.INT"

C1_AML027_DEG <- FindMarkers(AML.combined_2, ident.1 = "C1_AML027_post", ident.2 = "C1_AML027_pre", verbose = FALSE)
C1_AML027_DEG.sorted <- C1_AML027_DEG[order(C1_AML027_DEG$avg_logFC, decreasing = TRUE),]
head(C1_AML027_DEG.sorted, 8)
tail(C1_AML027_DEG.sorted, 8)

C1_AML035_DEG <- FindMarkers(AML.combined_2, ident.1 = "C1_AML035_post", ident.2 = "C1_AML035_pre", verbose = FALSE)
C1_AML035_DEG.sorted <- C1_AML035_DEG[order(C1_AML035_DEG$avg_logFC, decreasing = TRUE),]
head(C1_AML035_DEG.sorted, 8)
tail(C1_AML035_DEG.sorted, 8)
```

Visualization of differentially expressed genes in UMAP plots.

```{r C20, echo=TRUE, fig.height=4.5, fig.width=9, message=FALSE, warning=FALSE}
FeaturePlot(AML.combined, features = "HSPA5", split.by = "INT", max.cutoff = 3, cols = c("grey", "red"))
```

Visualization of differentially expressed genes in Vlnplots.

```{r C21, echo=TRUE, fig.height=4, fig.width=9, message=FALSE, warning=FALSE}
plots <- VlnPlot(AML.combined, features = "HSPA5", split.by = "INT", group.by = "celltype", pt.size = 0, combine = FALSE)
CombinePlots(plots = plots, ncol = 1)
```

Here is the output of sessionInfo() on the system on which this document was compiled:

```{r echo=TRUE}
sessionInfo()
```

